RandomizedSVD.C

```diff
@@ -0,0 +1,267 @@
+/******************************************************************************
+ *
+ * Copyright (c) 2013-2019, Lawrence Livermore National Security, LLC
+ * and other libROM project developers. See the top-level COPYRIGHT
+ * file for details.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR MIT)
+ *
+ *****************************************************************************/
+
+// Description: A class implementing interface of SVD for the Randomized SVD
+//              algorithm.
+
+#include "RandomizedSVD.h"
+
+#include "mpi.h"
+#include "scalapack_wrapper.h"
+
+#include <limits.h>
+#include <algorithm>
+#include <random>
+
+#include <stdio.h>
+#include <string.h>
+
+/* Use automatically detected Fortran name-mangling scheme */
+#define dormlq CAROM_FC_GLOBAL(dormlq, DORMLQ)
+
+extern "C" {
+void dormlq(char*, char*, int*, int*, int*,
+    double*, int*, double*, double*, int*, double*, int*, int*);
+}
+
+namespace CAROM {
+
+RandomizedSVD::RandomizedSVD(
+   Options options) :
+   StaticSVD(options),
+   d_subspace_dim(options.randomized_subspace_dim) {
+     srand(1);
+   }
+
+void
+RandomizedSVD::computeSVD()
+{
+     d_samples->n = d_num_samples;
+     delete_factorizer();
+
+     int num_rows = d_total_dim;
+     int num_cols = d_num_samples;
+     if (d_subspace_dim < 1 || d_subspace_dim > std::min(num_rows, num_cols)) {
+       d_subspace_dim = std::min(num_rows, num_cols);
+     }
+
+     // Get snapshot matrix in distributed format.
+     // If there are less dimensions than samples, use the transpose instead.
```

Comment:
Is there testing of both cases?

Reply:
Yes, currently there are 4 serial tests. One for full k, and one for smaller k for both cases (rows > cols and cols > rows)